reconcile/oum/base.py

```diff
@@ -0,0 +1,373 @@
+                ocm_env=ocm_env,
+                org_id=org_id,
+            ),
+            by=1 if len(errors) > 0 else 0,
+        )
+        metrics.set_gauge(
+            ValidationErrorsGauge(
+                integration=self.name,
+                ocm_env=ocm_env,
+                org_id=org_id,
+            ),
+            validation_errors,
+        )
+
+    @abstractmethod
+    def signal_cluster_reconcile_success(
+        self,
+        dry_run: bool,
+        ocm_api: OCMBaseClient,
+        spec: ClusterUserManagementSpec,
+        message: str,
+    ) -> None:
+        """
+        This method is called when the cluster reconcile operation was successful.
+        """
+
+    @abstractmethod
+    def signal_cluster_validation_error(
+        self,
+        dry_run: bool,
+        ocm_api: OCMBaseClient,
+        spec: ClusterUserManagementSpec,
+        error: Exception,
+    ) -> None:
+        """
+        This method is called when the configuration for a cluster is invalid.
+
+        If this method throws an exception, reconciliation on the rest of the clusters
+        of an organization will stop.
+        """
+
+    @abstractmethod
+    def signal_cluster_reconcile_error(
+        self,
+        dry_run: bool,
+        ocm_api: OCMBaseClient,
+        spec: ClusterUserManagementSpec,
+        error: Exception,
+    ) -> None:
+        """
+        This method is called when the cluster reconcile operation failed.
+
+        If this method throws an exception, reconciliation on the rest of the clusters
+        of an organization will stop.
+        """
+
+
+def reconcile_cluster_roles(
+    dry_run: bool,
+    ocm_api: OCMBaseClient,
+    org_id: str,
+    spec: ClusterUserManagementSpec,
+) -> ClusterRoleReconcileResult:
+    """
+    Make sure that the cluster roles and given users are synced onto the cluster.
+    This implies adding missing users and removing stale ones.
+    Cluster roles not mentioned in the spec are not touched.
+    """
+    result = ClusterRoleReconcileResult()
+
+    try:
+        org_id = spec.cluster.organization_id
+        cluster = spec.cluster.ocm_cluster
+        desired_groups: dict[OCMClusterGroupId, set[str]] = spec.roles
+        current_groups: dict[OCMClusterGroupId, set[str]] = {
+            role_id: {u.id for u in group.users.items}
+            for role_id, group in get_cluster_groups(
+                ocm_api=ocm_api, cluster_id=cluster.id
+            ).items()
+            if group.users
+        }
+        # only process roles present in the desired state and ignore the ones that are not
+        # this way a role can still be managed manually while another one is managed by this integration
+        for group in desired_groups.keys():
+            desired_members = desired_groups.get(group, set())
+            current_members = current_groups.get(group, set())
+
+            # add missing users
+            for missing_user in desired_members - current_members:
```

Comment:
nit: Do you want to try the generic [differ](https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/differ.py) here?

to do group level diff
```python
diff_result = diff_mappings(current_groups, desired_groups)
```

To do member level diff

```python
diff_result = diff_iterables(current_members, desired_members)
```

then process different parts

```python
diff_result.add
diff_result.delete
```

Since no need to delete current groups, it's not that useful, just save some mind burden of calculating diff.

Reply:
oh i forgott about the generif differ. wanted to try that one.
i will update the PR

Reply:
after having a closer look i don't think it helps a lot here because diffing is really just a `-` operation beetween to sets of strings. i will keep the differ in mind though for other scenarios where diffs are more cumbersome to calculate.

ok with you?

Reply:
Yes, diff here is very simple, generic differ is more suitable where we need to find things to add, delete, change and not changed.